src/framework/theme/components/menu/menu.service.ts

```diff
@@ -281,11 +293,21 @@ export class NbMenuInternalService {
   private selectedInUrl(item: NbMenuItem): boolean {
     const exact: boolean = item.pathMatch === 'full';
     const location: string = this.location.path();
+    const locationPath: string = this.location.path().match(/.*?(?=[?#]|$)/)[0];
+    const locationFragment: string = location.indexOf('#') !== -1 ? location.substr(location.indexOf('#') + 1) : null;
+    const endOfUrlSegmentRegExp = /\/|^$/;
 
-    return (
-      (exact && location === item.link) ||
-      (!exact && location.includes(item.link)) ||
-      (exact && item.fragment && location.substr(location.indexOf('#') + 1).includes(item.fragment))
+    const selected = (
```

Comment:
how about we divide this logic into smaller helpers or even a separate helper class? and create a unit test for it.